Documentation/rust/quick-start.rst

```diff
@@ -3,97 +3,132 @@
 Quick Start
 ===========
 
-This document describes how to get started with Rust kernel development.
+This document describes how to get started with kernel development in Rust.
 
 
 Requirements
 ------------
 
-rustc and cargo
-***************
+This section explains how to fetch the requirements to work with Rust.
+If you have worked previously with Rust, this will only take a moment.
 
-A recent nightly Rust toolchain (with, at least, ``rustc`` and ``cargo``)
-is required, e.g. ``nightly-2021-01-02``. In the future, this restriction
-will be lifted.
+Some of these requirements might be available from your Linux distribution
+under names like ``rustc``, ``rust-src``, ``rust-bindgen``, etc. However,
+at the time of writing, they are likely to not be recent enough.
+
+
+rustc
+*****
+
+A recent *nightly* Rust toolchain (with, at least, ``rustc``) is required,
+e.g. ``nightly-2021-01-02``. Our goal is to use a stable toolchain as soon
+as possible, but for the moment we depend on a handful of nightly features.
 
 If you are using ``rustup``, run::
 
     rustup toolchain install nightly
 
-Otherwise, fetch a standalone installer from:
+Otherwise, fetch a standalone installer or install ``rustup`` from:
 
     https://www.rust-lang.org
 
 
-rustc sources
-*************
+Rust standard library source
+****************************
 
-The sources for the compiler are required to be available because the standard
-library (``core`` and ``alloc``) is cross-compiled.
+The Rust standard library source (``core`` and ``alloc``, at least) is required
+because the build system will cross-compile it.
 
 If you are using ``rustup``, run::
 
     rustup component add rust-src
 
-Otherwise, if you used a standalone installer, you can clone the Rust compiler
-sources and create a symlink to them in the installation folder of
-your nightly toolchain::
+Otherwise, if you used a standalone installer, you can clone the Rust
+repository into the installation folder of your nightly toolchain::
+
+    git clone https://github.com/rust-lang/rust `rustc --print sysroot`/lib/rustlib/src/rust
+
+
+compiler-builtins source
+************************
 
-    git clone https://github.com/rust-lang/rust
-    ln -s rust .../rust-nightly/lib/rustlib/src
+The source for ``compiler-builtins`` (a Rust port of LLVM's ``compiler-rt``)
+is required.
+
+The build system expects the sources alongside the Rust ones we just installed,
+so you can clone it into the installation folder of your nightly toolchain::
+
+    git clone https://github.com/rust-lang/compiler-builtins `rustc --print sysroot`/lib/rustlib/src/compiler-builtins
```

Comment:
What would it take to make `compiler-builtins` be something that didn't need to manually be installed? For rustup to include it?

Reply:
If `rustup` includes it in the default profile, yeah, but I don't think it is useful for the vast majority of Rust users, so it is hard to argue for them to include it. Nevertheless, it would be useful to have it as a component in rustup for easy installation. Even better, the `rust` repo could include it, since `alloc` requires it (at least for some targets); but I am not sure why they didn't include it.

Having said that: I'd not worry too much for the moment, since we don't know if we will use it. For instance, we could simply include the builtins needed (which is what the kernel does for a few, actually). I did it this way for the moment since it is the "normal" way -- I haven't taken a look at removing it yet.

So I suggest filling an issue to keep track of it.

Reply:
#69

Reply:
Would it make sense to use a submodule instead? Or download/check in the `.crate` file from crates.io? The `.crate` files are about 150k:

```
-rw-r--r-- 1 bjorn bjorn 134K dec 26 12:48 /home/bjorn/.cargo/registry/cache/github.com-1ecc6299db9ec823/compiler_builtins-0.1.32.crate
-rw-r--r-- 1 bjorn bjorn 152K dec  1 22:01 /home/bjorn/.cargo/registry/cache/github.com-1ecc6299db9ec823/compiler_builtins-0.1.36.crate
-rw-r--r-- 1 bjorn bjorn 155K dec 17 18:26 /home/bjorn/.cargo/registry/cache/github.com-1ecc6299db9ec823/compiler_builtins-0.1.37.crate
-rw-r--r-- 1 bjorn bjorn 155K dec 17 16:46 /home/bjorn/.cargo/registry/cache/github.com-1ecc6299db9ec823/compiler_builtins-0.1.38.crate
```

Reply:
Submodules could be a way, but if we were to go that route, then at that point it is cleaner to import the files instead, possibly keeping only the needed subset as mentioned above. As for `.crate` files, aren't they source tarballs? What would be the advantage vs. checking out the repo in that case?

Reply:
It would also be possible to just vendor the crate. The advantage over checking out the repo is that it will make it unnecessary to clone a repo outside of the control of both distros and the Linux foundation.

Reply:
> It would also be possible to just vendor the crate.

Do you mean `cargo vendor`ing? That was suggested in the past, but I don't see why we would want to use cargo just for that. What do you mean?

>  The advantage over checking out the repo is that it will make it unnecessary to clone a repo outside of the control of both distros and the Linux foundation.

I was referring to the "download from crates.io" option you were discussing, not the "checking in" one. (Anyway, the repo or the crate files could be in kernel.org, that would not be a problem).

Reply:
Indeed

Reply:
Got it -- updating shouldn't be a problem whether is one file or many, in principle. I feel having a tarball checked in wouldn't be very welcome in the kernel -- but I may be wrong, and it is an option to keep on the table nevertheless.